optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/ConstraintTree.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Stack;
+
+import org.optaplanner.core.impl.score.stream.drools.common.consequences.ConstraintConsequence;
+import org.optaplanner.core.impl.score.stream.drools.common.nodes.ChildNode;
+import org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNode;
+import org.optaplanner.core.impl.score.stream.drools.common.nodes.ConstraintGraphNodeType;
+
+final class ConstraintTree<Node_ extends ConstraintGraphNode, Consequence_ extends ConstraintConsequence<Node_>> {
+
+    private final ConstraintSubTree nestedNodes;
+
+    ConstraintTree(Consequence_ consequence) {
+        List<ConstraintGraphNode> orderedNodeList = orderNodes(consequence);
+        this.nestedNodes = assembleSubTree(orderedNodeList);
+        ConstraintGraphNode firstNode = orderedNodeList.get(0);
+        if (firstNode.getType() != ConstraintGraphNodeType.FROM) {
+            throw new IllegalStateException("First node (" + firstNode + ") is not a " + ConstraintGraphNodeType.FROM +
+                    " (" + firstNode.getType() + ").");
+        }
+        ConstraintGraphNode lastNode = orderedNodeList.get(orderedNodeList.size() - 1);
+        if (lastNode != consequence.getTerminalNode()) {
+            throw new IllegalStateException("Last node (" + lastNode + ") is not the terminal node (" +
+                    consequence.getTerminalNode() + ").");
+        }
+    }
+
+    private List<ConstraintGraphNode> orderNodes(Consequence_ consequence) {
+        // Depth-first search, right parents of join nodes get precedence.
+        List<ConstraintGraphNode> nodeList = new ArrayList<>(0);
+        Stack<ConstraintGraphNode> unprocessedNodes = new Stack<>();
```

Comment:
Stack is a relic from Java 1.0, just like Vector.

Vector was replaced by (Array)List
Stack was replaced by (Array)Deque